### ISSUE TITLE
RoBERTa on SuperGLUE's BoolQ task

### ISSUE THREAD

**Issue by dirkgr** [allenai/allennlp#4996]
BoolQ is one of the tasks of the [SuperGLUE](https://super.gluebenchmark.com) benchmark. The task is to re-trace the steps of Facebook's RoBERTa paper (https://arxiv.org/pdf/1907.11692.pdf) and build an AllenNLP config that reads the BoolQ data and fine-tunes a model on it. We expect scores in the range of their entry on the [SuperGLUE leaderboard](https://super.gluebenchmark.com/leaderboard).

We recommend you use the [AllenNLP Repository Template](https://github.com/allenai/allennlp-template-config-files) as a starting point. It might also be helpful to look at how the TransformerQA model works ([training config](https://github.com/allenai/allennlp-models/blob/main/training_config/rc/transformer_qa.jsonnet), [reader](https://github.com/allenai/allennlp-models/blob/main/allennlp_models/rc/dataset_readers/transformer_squad.py), [model](https://github.com/allenai/allennlp-models/blob/main/allennlp_models/rc/models/transformer_qa.py)).

**Comment by wangyuxinwhy** [comment]
I'm very interested in this contribution and I have 2 years experience using allennlp.But there is one thing to be mentioned, I am from China.I'm afraid that my poor English skills will limit it, but I will try my best.Do you mind？

Now I have two questions to clarify

1.  use allennlp 2.0 api ?
2. I create a project in my own repositories that fully runs the BoolQ task and show it to you? Or fork allennlp and have a PR?

**Comment by dirkgr** [comment]
No problem about your English skills. Judging by this message, you're doing alright :-)

1. Yes please, these should work with the latest version.
2. Please create a project in your own repository. We have templates ([with config files](https://github.com/allenai/allennlp-template-config-files) and [with the API directly](https://github.com/allenai/allennlp-template-python-script)), and if you are very familiar with AllenNLP already, you can start from scratch.

If the model ends up very good, we might take it into [allennlp-models](https://github.com/allenai/allennlp-models), but that comes later.

Note that we already have a multiple-choice model at https://github.com/allenai/allennlp-models/blob/main/allennlp_models/mc/models/transformer_mc.py. This one could be very similar. Maybe you can save a lot of time by stealing code from there!

**Comment by wangyuxinwhy** [comment]
OK, let's do it.I find two related statements about "BoolQ & RoBERTa".

1. from fairseq issue, https://github.com/pytorch/fairseq/issues/1027 

> The input format we used for BoolQ was same as other tasks ie. <s> passage </s> </s> question </s> and we feed first <s> representation to the classifier. While truncation for seq_len>512, we always truncate passage.

2. from allenai EMNLP'20, https://arxiv.org/pdf/2004.04849.pdf

> We train the model on two-way questions using the input format: “[CLS] passage [SEP] question [SEP] answer”. The model scores each answer (“yes” or “no”) by applying a linear classifier over the [CLS] representation for each answer’s corresponding input.

The first , BoolQ as a binary classification task.
The second, BoolQ as a MC task.

which way should i follow? what’s your opinion?

**Comment by dirkgr** [comment]
They are both binary classification tasks. The MC model works by feeding in one sequence for every answer choice, for example, "[cls] question [sep] answer1", "[cls] question [sep] answer2", and so on. Then they compute a score for each answer choice based on the [cls]. I don't think that would work very well in this case, since the answers would only be "yes" or "no".

So it makes more sense to just get a score from the [cls], and then say the answer is "yes" if the score is positive, and "no" if the score is negative, or something like that. In other words, let's treat it as binary classification.

**Comment by wangyuxinwhy** [comment]
Thank you for your advice and support. I will do it as soon as possible. 🤗

**Comment by wangyuxinwhy** [comment]
Great, I would love to make a PR to allennlp-models.

**Comment by wangyuxinwhy** [comment]
@dirkgr Maybe we can close this issue.

I am almost familiar with how to participate in the development of allenNLP models, and I hope to do more for allenNLP. For example, I can implement all the models related to SuperGlue, and I can also help AllenNLP with Chinese translation. 

Do you have any suggestions for me to participate in the construction of AllenNLP community?

**Comment by dirkgr** [comment]
Yes, I think this issue is done. Thank you for your work! Do you mind if we credit you on https://gallery.allennlp.org? I'm thinking about making a page there for the project of SuperGLUE on AllenNLP.

We have done some work trying to find out what's most important to researchers, and we find that the most important thing is always baselines. Not having to implement somebody else's model saves researchers a lot of time. So the biggest thing you could for for AllenNLP is implement more baselines. All the tasks in SuperGLUE would be a great start, and probably pretty easy. There is also the rest of the tasks in https://api.semanticscholar.org/CorpusID:208637516, or really any of the issues under the [Models](https://github.com/allenai/allennlp/labels/Models) label. If none of those are interesting to you, write me and I'll find something better. I have a list :-)

I promise I'll be faster to respond, too. I got a little busy with getting https://gallery.allennlp.org up and running.

**Comment by dirkgr** [comment]
Oh, I forgot to answer the question about Chinese. I don't know how interesting Chinese translation is for AllenNLP users, since most of the models themselves only work on English text. Is it useful to have Chinese documentation for models that do question answering in English?

It would be cool though to re-train a bunch of the models to work in other languages. I don't know much about that space myself. What datasets are available? If you want to take on a big project like "make AllenNLP work in Chinese", we'll help you in whatever way we can.

**Comment by wangyuxinwhy** [comment]
@dirkgr It is my honor to join https://gallery.allennlp.org , I also think SuperGlue is a good place to start, and I will implement all the baseline models it includes in the near future. I will try to make a preliminary version first, and then I need your help to review it and see what needs to be improved.

I have used allenNLP as the main framework in many Chinese and English NLP competitions, and achieved good results. From my perspective, it is usually because of the limitation of Chinese documents and learning materials that people do not use allenNLP more widely in China.It's really a pity!

I just want to combine my strengths to do something for AllenNLP. Chinese is my native language, and I'm familiar with the source code of AllenNLP, so combining the two was a natural choice, but I'm not sure where to go. I can try a few things first, see what the response is, and then ask for your advice and help when I have something in mind.

**Comment by dirkgr** [comment]
I think the main entry point for getting started with AllenNLP is the guide at http://guide.allennlp.org. If you want to translate the guide into Chinese, we'll figure out how to host it. I'd say it would go on the same site, and we'll put little flags in the corner for every page so people can switch language?
